src/core/lib/event_engine/ares_driver.cc

```diff
@@ -0,0 +1,952 @@
+      return node_ == other.node_;
+    }
+    FdNodeListIterator& operator++(int) {
+      node_ = node_->next;
+      return *this;
+    }
+    FdNode* operator*() { return node_; }
+    static FdNodeListIterator universal_end() {
+      return FdNodeListIterator(nullptr);
+    }
+
+   private:
+    friend class FdNodeList;
+    explicit FdNodeListIterator(FdNode* node) : node_(node) {}
+    FdNode* node_;
+  };
+  using Iterator = FdNodeListIterator;
+
+  FdNodeList() = default;
+  ~FdNodeList() { GPR_ASSERT(IsEmpty()); }
+
+  Iterator begin() { return Iterator(head_); }
+  Iterator end() { return Iterator::universal_end(); }
+
+  bool IsEmpty() const { return head_ == nullptr; }
+
+  void PushFdNode(FdNode* fd_node)
+      ABSL_EXCLUSIVE_LOCKS_REQUIRED(&GrpcAresRequest::mu_) {
+    fd_node->next = head_;
+    head_ = fd_node;
+  }
+
+  FdNode* PopFdNode() ABSL_EXCLUSIVE_LOCKS_REQUIRED(&GrpcAresRequest::mu_) {
+    GPR_ASSERT(!IsEmpty());
+    FdNode* ret = head_;
+    head_ = head_->next;
+    return ret;
+  }
+
+  // Search for as in the FdNode list. This is an O(n) search, the max
+  // possible value of n is ARES_GETSOCK_MAXNUM (16). n is typically 1 - 2
+  // in our tests.
+  FdNode* PopFdNode(ares_socket_t as)
+      ABSL_EXCLUSIVE_LOCKS_REQUIRED(&GrpcAresRequest::mu_) {
+    FdNode phony_head;
+    phony_head.next = head_;
+    FdNode* node = &phony_head;
+    while (node->next != nullptr) {
+      if (node->next->polled_fd->GetWrappedAresSocketLocked() == as) {
+        FdNode* ret = node->next;
+        node->next = node->next->next;
+        head_ = phony_head.next;
+        return ret;
+      }
+      node = node->next;
+    }
+    return nullptr;
+  }
+
+ private:
+  FdNode* head_ = nullptr;
+};
+
+GrpcAresRequest::GrpcAresRequest(
+    absl::string_view name, absl::optional<absl::string_view> default_port,
+    EventEngine::Duration timeout,
+    RegisterAresSocketWithPollerCallback register_cb, EventEngine* event_engine)
+    : grpc_core::RefCounted<GrpcAresRequest>(
+          GRPC_TRACE_FLAG_ENABLED(grpc_trace_ares_driver) ? "GrpcAresRequest"
+                                                          : nullptr),
+      name_(name),
+      default_port_(default_port.has_value() ? *default_port : ""),
+      timeout_(timeout),
+      fd_node_list_(std::make_unique<FdNodeList>()),
+      event_engine_(event_engine),
+      polled_fd_factory_(NewGrpcPolledFdFactory(std::move(register_cb), &mu_)) {
+}
+
+GrpcAresRequest::~GrpcAresRequest() {
+  if (initialized_) {
+    ares_destroy(channel_);
+    GRPC_ARES_DRIVER_TRACE_LOG("request:%p destructor", this);
+    GRPC_ARES_DRIVER_STACK_TRACE();
+  }
+}
+
+absl::Status GrpcAresRequest::Initialize(absl::string_view dns_server,
+                                         bool check_port) {
+  absl::MutexLock lock(&mu_);
+  GPR_DEBUG_ASSERT(!initialized_);
+  absl::string_view port;
+  GPR_ASSERT(grpc_core::SplitHostPort(name_, &host_, &port));
```

Comment:
I believe the assert is wrong here, it was crashing on the naming resolver tests when I patched your PR in.